Add PGS to SRT OCR conversion feature

mikeSGman opened the pull request:
# Add PGS to SRT OCR Conversion Feature

## Summary

This PR adds support for converting image-based PGS (Presentation Graphic Stream) subtitles to text-based SRT format using OCR (Optical Character Recognition). This feature enables users to extract Blu-ray subtitles as editable text files.

## Motivation

PGS subtitles are image-based and cannot be edited or searched. Many users want to:
- Edit subtitle timing or text content
- Search subtitle content
- Reduce file sizes (text SRT vs. image SUP)
- Use subtitles with devices/players that don't support PGS

## Features

### User-Facing Changes

1. **Dropdown Menu for PGS Subtitles**
   - PGS subtitle tracks now show a dropdown menu with two options:
     - "Extract as .sup (image - fast)" - Instant extraction of image-based subtitles
     - "Convert to .srt (OCR - 3-5 min)" - OCR conversion to text-based subtitles

2. **Settings Panel**
   - New checkbox: "Enable PGS to SRT OCR conversion"
   - Dependency status display showing which OCR tools are installed
   - Link to installation instructions for missing dependencies

3. **Smart Dependency Detection**
   - Auto-detects Tesseract OCR on all drives (C:, D:, E:, etc.)
   - Checks Windows registry for Tesseract installation
   - Auto-detects MKVToolNix (mkvmerge)
   - Auto-detects pgsrip Python package
   - Gracefully falls back to .sup extraction if dependencies are missing

4. **User-Friendly Error Messages**
   - Clear instructions for installing missing dependencies
   - Platform-specific installation commands (Windows, Linux, macOS)
   - Links to download pages for manual installation

### Technical Implementation

#### Files Modified

1. **fastflix/models/config.py**
   - Added `find_ocr_tool()` function to locate Tesseract, mkvmerge, and pgsrip
   - Searches system PATH, environment variables, common install locations, and Windows registry
   - Added config fields: `enable_pgs_ocr`, `tesseract_path`, `mkvmerge_path`, `pgsrip_path`

2. **fastflix/widgets/background_tasks.py**
   - Added `_check_pgsrip_dependencies()` method to verify all required tools
   - Added `_convert_sup_to_srt()` method to perform OCR conversion
   - Handles language code conversion (ISO 639-2/T 3-letter → ISO 639-1 2-letter)
   - Sets environment variables (PATH and TESSERACT_CMD) for pytesseract
   - Added `use_ocr` parameter to `ExtractSubtitleSRT` class

3. **fastflix/widgets/panels/subtitle_panel.py**
   - Modified the Extract button for PGS tracks to show a dropdown menu
   - Conditionally enables OCR option based on settings and dependencies
   - Shows helpful tooltips when OCR option is disabled

4. **fastflix/widgets/settings.py**
   - Added PGS OCR settings checkbox with tooltip
   - Added `update_ocr_dependency_status()` method to show dependency status
   - Displays checkmarks for installed dependencies
   - Shows link to wiki for installation instructions when dependencies are missing

5. **FastFlix_Windows_OneFile.spec**
   - Added pgsrip and its dependencies to the PyInstaller build
   - Ensures OCR libraries are bundled in compiled Windows executable

#### Dependencies

- **Tesseract OCR** - OCR engine for text recognition
- **MKVToolNix (mkvmerge)** - Required by pgsrip for subtitle extraction
- **pgsrip** - Python library for PGS subtitle OCR conversion
  - Automatically installs: pytesseract, opencv-python, numpy, pysrt, babelfish, cleanit

#### Key Design Decisions

1. **Two-Step Process**: First extract .sup file using FFmpeg, then convert with pgsrip
   - Separates FFmpeg operations from OCR operations
   - Allows fallback to .sup extraction if OCR fails
   - Provides better error handling

2. **Language Code Conversion**: Automatically converts ISO 639-2/T (eng) to ISO 639-1 (en)
   - pgsrip expects 2-letter language codes in filenames
   - Maintains compatibility with FastFlix's 3-letter language codes

3. **Environment Variable Management**: Sets both TESSERACT_CMD and PATH
   - TESSERACT_CMD points to tesseract.exe
   - PATH includes Tesseract directory for subprocess calls
   - Fixes issue where pytesseract can't find Tesseract on Windows

4. **Automatic Cleanup**: Deletes .sup file after successful .srt conversion
   - Keeps only the text-based .srt file
   - Reduces clutter in the output directory

## Testing

### Tested Scenarios

- ✅ Windows with Tesseract on D: drive (non-standard location)
- ✅ PGS subtitle extraction and OCR conversion
- ✅ Language code conversion (eng → en)
- ✅ Dependency detection and status display
- ✅ Settings UI enable/disable functionality
- ✅ Dropdown menu for PGS tracks
- ✅ Fallback to .sup when dependencies are missing
- ✅ Error handling and user-friendly messages

### Test Results

- Successfully converted Blade (1998) Blu-ray PGS subtitles to SRT
- Conversion time: ~46 seconds for a feature film
- Output: Clean .srt file with proper timing

### Testing Checklist

- [x] Test on Windows with Tesseract in the default location (C:\Program Files)
- [x] Test on Linux with apt-installed dependencies
- [ ] Test on macOS with Homebrew-installed dependencies
- [ ] Test with Spanish, French, and other language subtitles
- [x] Test with missing dependencies (verify error messages)
- [x] Test dropdown menu behavior with OCR enabled/disabled
- [x] Test PyInstaller build includes pgsrip dependencies
- [x] Verify settings persist after restart

## Installation Instructions for Users

### Windows

```cmd
# Install Tesseract OCR
# Download installer from: https://github.com/UB-Mannheim/tesseract/wiki

# Install MKVToolNix
# Download installer from: https://mkvtoolnix.download/downloads.html

# Install pgsrip (from FastFlix virtual environment)
pip install pgsrip
```

### Linux

```bash
sudo apt install tesseract-ocr mkvtoolnix
pip install pgsrip
```

### macOS

```bash
brew install tesseract mkvtoolnix
pip install pgsrip
```

## Breaking Changes

None. This is a purely additive feature that's disabled by default.

## Migration Guide

No migration needed. Existing users will see the new option after updating and installing dependencies.

## Future Enhancements

Potential improvements for future PRs:
- Support for more OCR languages via Tesseract language packs
- Batch conversion of multiple PGS tracks
- OCR accuracy tuning options
- Progress bar for long OCR operations
- Integration with online OCR services as a fallback

## Related Issues

Closes #[issue-number] (if applicable)

## Screenshots

<img width="1286" height="835" alt="image" src="https://github.com/user-attachments/assets/93b76d54-78c4-42e2-b2a8-0cc8faf0c8fd" />

<img width="1100" height="499" alt="image" src="https://github.com/user-attachments/assets/9cc4e0ff-2e0d-4c04-85d7-5cc84304197c" />

## Checklist

- [x] Code follows project style guidelines
- [x] Comments added for complex logic
- [x] No debug/console.log statements
- [x] User-facing strings use translation function `t()`
- [x] Error handling for all external tool calls
- [x] Graceful degradation when dependencies are missing
- [x] Platform-specific code tested on Windows
- [x] Config fields have sensible defaults
- [x] Feature is opt-in (disabled by default)

cdgriffith commented on fastflix/widgets/background_tasks.py:
```diff
@@ -164,6 +178,132 @@ def _get_subtitle_format(self):
+            self.main.thread_logging_signal.emit(
+                f"ERROR:{t('Missing dependencies for PGS OCR')}: {', '.join(missing)}\n\n"
+                f"Install instructions:\n"
+                f"  Windows: Run setup_pgs_ocr_windows.bat in FastFlix folder\n"
```
This file does not exist in the PR

cdgriffith reviewed (commented):
Thank you so much for this wonderful addition! 

I have a few tweaks suggested. If you are up for doing them let me know, otherwise I can merge it and work on it as well as you set up a great feature I would love to add! 

If you do add more please also run the pre-commit checks so it passes linting: 

```
pre-commit install
pre-commit run --all-files
```

cdgriffith commented on fastflix/models/config.py:
```diff
@@ -99,6 +99,77 @@ def where(filename: str, portable_mode=False) -> Path | None:
+            ]
+            for path in common_paths:
+                if path.exists():
+                    return path
```
I would keep the logic in to look on the path if it fails to find them in these locations, but wouldn't just start searching on random hard drives and use designated system paths.

cdgriffith commented on fastflix/models/config.py:
```diff
@@ -168,6 +239,13 @@ class Config(BaseModel):
+    enable_pgs_ocr: bool = False
+    tesseract_path: Path | None = Field(default_factory=lambda: find_ocr_tool("tesseract"))
+    mkvmerge_path: Path | None = Field(default_factory=lambda: find_ocr_tool("mkvmerge"))
+    pgsrip_path: Path | None = Field(default_factory=lambda: find_ocr_tool("pgsrip"))
```
If there were issues with that way let me know! I just assumed would be more reliable and easier to work in the code that way.

mikeSGman replied:
done.

mikeSGman commented:
Hi @cdgriffith - My earlier commit missed the `BabelLanguage` patch for dual 2- and 3-letter ISO code handling. This push includes that fix; verified locally, and all pre-commit hooks are passing. 

Based on our threaded convo, I think I've addressed all issues:
```
    - Use environment variables for Windows tool detection instead of
      scanning all drives (LOCALAPPDATA, PROGRAMFILES, PROGRAMFILES(X86))
    - Remove pgsrip_path config field and use pgsrip Python API directly
    - Update dependency checks to use importlib for pgsrip library
    - Fix BabelLanguage to handle both 2-letter and 3-letter ISO codes
    - Update error messages and installation instructions
```
<img width="1460" height="1059" alt="image" src="https://github.com/user-attachments/assets/ec878763-ec3d-47c7-ab03-336cc2add8fa" />

mikeSGman commented:
I give up, I simply can't figure out how to get it to work on a compiled binary (it compiles cleanly - but the srt extraction fails). It works perfectly from source, though, so that's good enough for my usecase.

mikeSGman commented:
I tried to, but it says: 
<img width="598" height="143" alt="image" src="https://github.com/user-attachments/assets/c3b24e9e-418f-448c-b5da-0291c7748b91" />
